server/storage/backend/batch_tx.go

```diff
@@ -53,6 +53,15 @@ type BatchTx interface {
 	Commit()
 	// CommitAndStop commits the previous tx and does not create a new one.
 	CommitAndStop()
+
+	// LockWithoutHook doesn't execute the `txPostLockHook`, while the Lock method may
+	// call back the hook if present.
+	//
+	// The original Lock() is supposed to be called only by operations in the applying
+	// workflow, and all other operations should call LockWithoutHook(). If the operation
+	// doesn't have any impact on the applying workflow, such as the `etcdutl` commands,
+	// then it doesn't matter which lock method it calls.
+	LockWithoutHook()
```

Comment:
Having the preferred locking method called `LockWithoutHooks` is just asking for trouble. Based on the comment looks like `LockWithoutHooks` should be used in most of the code and `Lock` only in one place apply logic. 

Can we find a way to split transaction interface, so that `Lock` in apply would use hooks and `Lock` everywhere else will not use hooks?

Reply:
Thanks @serathius for the feedback. It's a valid comment, and @ptabor raised similar concern above. 

Firstly, if you go through the apply workflow, you will find that the users of `Lock` in the apply workflow overnumbers the users of `LockWithoutHooks` in other places.

Secondly, I have been thinking all the possible solutions (see below) in the past week, and eventually the solution in this PR is the simplest one, accordingly the least impact. 

All the possible solutions I can think of:
1. add a LockWithoutHook, which is supposed to be called outside the apply workflow. It's exactly what this PR delivers. It's the simplest solution and the least impact. The bad side is it's error prone, because it may be a little difficult to main the correct calls just as @ptabor pointed out.
2. add a `backend` wrapper (originally proposed by @ptabor ) and a `batchTxBuffered` wrapper; it just adds an additional functionality hooking on top of the existing design/implementation. Only the apply workflow holds a reference to the wrapped backend, all others keep unchanged.  It's the most decent solution so far. But the bad side is it needs big refactor the repo. See the concept in [pull/17](https://github.com/ahrtr/etcd/pull/17)
3. to be done (need to have breakfast for now)